modules/ITK/cli/utils/statismo-build-gp-model-kernels.h

```diff
@@ -95,12 +95,181 @@ class GaussianKernel : public statismo::ScalarValuedKernel<TPoint> {
+            kUpper[d] = static_cast<int>(std::floor(std::min(xScaled[d], yScaled[d]) + 0.5 * supportBasisFunction));
+        }
+
+
+        // We need to generate the cartesian product k_1 x ... x k_d, where k_i goes through all the integers
+        // within the given bounds. A non-recursive solution requires d loops. Here we just write down the cases
+        // for 1 2 and 3D
+        double sum = 0.0;
+        double kx = kLower[0];
+        while (kx <= kUpper[0]) {
+            if (dim == 1) {
+                VectorType k(1);
+                k[0] = kx;
+                sum += (tensorProductSpline(xScaled - k) * tensorProductSpline(yScaled - k));
+
+            } else {
+                double ky = kLower[1];
+                while (ky <= kUpper[1]) {
+                    if (dim == 2) {
+                        VectorType k(2);
+                        k[0] = kx; k[1] = ky;
+                        sum += (tensorProductSpline(xScaled - k) * tensorProductSpline(yScaled - k));
+                    } else {
+                        double kz = kLower[2];
+                        while (kz <= kUpper[2]) {
+                            VectorType k(3);
+                            k[0] = kx; k[1] = ky; k[2] = kz;
+                            sum += (tensorProductSpline(xScaled - k) * tensorProductSpline(yScaled - k));
+                            kz += 1;
+                        }
+                    }
+                    ky += 1;
+                }
+            }
+
+            kx += 1;
+        }
+
+        return sum;
+    }
+
+    std::string GetKernelInfo() const {
+        std::ostringstream os;
+        os << "BSplineKernel(" << m_support << ")";
+        return os.str();
+    }
+
+private:
+    double m_support;
+};
+
+
+//
+// A kernel that represents deformations on different scale levels.
+// Each scale level is modeled as a b-spline kernel, which leads to a "wavelet style"
+// decomposition of the deformation. See
+// Opfer, Roland. "Multiscale kernels." Advances in computational mathematics 25.4 (2006): 357-380.
+//
+template <class TPoint>
+class MultiscaleKernel : public statismo::ScalarValuedKernel<TPoint> {
+public:
+    typedef typename  TPoint::CoordRepType CoordRepType;
+    typedef vnl_vector<CoordRepType> VectorType;
+
+
+    MultiscaleKernel(double supportBaseLevel, unsigned numberOfLevels) :
+            m_supportBaseLevel(supportBaseLevel),
+            m_numberOfLevels(numberOfLevels)
+    {
+        double support = supportBaseLevel;
+        for (unsigned i = 0; i < numberOfLevels; ++i) {
+            m_kernels.push_back(new BSplineKernel<TPoint>(m_supportBaseLevel * std::pow(2, -1.0 * i)));
+        }
+    }
+
+
+
+    inline double operator()(const TPoint& x, const TPoint& y) const {
+
+        assert(x.Size() == y.Size());
+        const unsigned dim = x.Size();
+        double sum = 0;
+        for (unsigned i = 0; i < m_kernels.size(); ++i) {
+            sum += m_kernels[i](x, y) * std::pow(2, -1.0 * i);
```

Comment:
do you really need to set once again std::pow( 2, -1.0 * i )? can't you skip this part or avoid doing it for all evaluations?